File: runtimes/swiftui/Sources/StyleConverterRuntime/Renderer/VerticalTextFlowLayout.swift
//
//  VerticalTextFlowLayout.swift
//  StyleConverterRuntime/Renderer — wave 35 lane B5, the renderer's
//  VERTICAL-FLOW region.
//
//  The SwiftUI twin of
//  runtimes/compose/src/main/java/com/styleconverter/runtime/core/renderer/
//  VerticalTextFlowLayout.kt. One view, owned by `writing-mode` and nothing
//  else, and every decision it takes comes from the pure module
//  `StyleEngine/typography/writing/VerticalTextFlow.swift` — so the Kotlin and
//  Swift renderers cannot drift about which runs are upright or where a
//  vertical line breaks.
//
//  ## What it draws
//  An UPRIGHT vertical run: glyphs stand up and stack DOWN each line, lines
//  stack across the block axis (right-to-left for `vertical-rl`, left-to-right
//  for `vertical-lr` — css-writing-modes-4 §3).
//
//  ## Why iOS needs it
//  Measured on the frozen wave34-final capture of WPT
//  css-writing-modes/available-size-011 (fullwidth "ＰＡＳＳ",
//  Vertical_Orientation U ⇒ upright under the initial `text-orientation:
//  mixed`): iOS has no vertical text path at all and paints the run
//  horizontally in SOURCE order — "Ｓ Ｓ Ａ Ｐ" where the browser-ref reads
//  "ＰＡＳＳ", 0.9415 against a web 0.9898. The letters are right; only the
//  line stacking is missing, and line stacking is exactly what this layout is.
//
//  ## The decline contract, and the one honest gap
//  The wrap budget is the block-axis extent available to the run, which
//  SwiftUI only reveals as `proposal.height` at measure time. When it is
//  unspecified — and `PlaceholderLabel`'s chain ends in
//  `.fixedSize(horizontal: false, vertical: true)`, which proposes nil on
//  exactly that axis — `VerticalTextFlow.uprightColumnIndices` DECLINES and
//  this layout places the `fallback` subview instead, i.e. the frozen
//  horizontal label, byte for byte. `budgetOverridePx` exists for the caller
//  that can name the extent from the IR rather than the proposal; nil means
//  "ask the proposal", which is all any caller does today.
//

import SwiftUI

/// THE GATE: does this run take the upright vertical path, and if so which way
/// do its lines stack? `nil` = "no", and the caller must keep whatever it did
/// before — a horizontal mode, a sideways mode, an ASCII run under `mixed`, an
/// explicit `text-orientation: sideways`, or a run that needs both classes.
///
/// One entry point so the decision has ONE spelling per platform; the Kotlin
/// twin is `verticalUprightStack(config:text:)` in
/// core/renderer/VerticalTextFlowLayout.kt, which reaches the same two
/// `VerticalTextFlow` calls through `TextExtractor.extractWritingModeConfig`.
enum VerticalUprightGate {
    /// - Parameters:
    ///   - properties: the MERGED, inheritance-resolved list — `writing-mode`
    ///     and `text-orientation` are both `Inherited: yes`
    ///     (css-writing-modes-4 §3.1 / §5.1), so they usually sit on an
    ///     ancestor and the component's own list would answer "horizontal".
    ///   - text: the run as it will be painted (post text-transform).
    static func stack(properties: [IRProperty], text: String?) -> LineStack? {
        guard let text, !text.isEmpty else { return nil }
        // No `writing-mode` anywhere in the chain ⇒ horizontal ⇒ not ours.
        guard let mode = WritingModeExtractor.extract(from: properties)?.mode else { return nil }
        let orientation = VerticalTextFlow.runOrientation(
            writingMode: mode,
            // Absent `text-orientation` ⇒ `.mixed`, the initial value.
            textOrientation: TextOrientationExtractor.extract(from: properties)?.value ?? .mixed,
            text: text)
        guard orientation == .upright else { return nil }
        return VerticalTextFlow.lineStack(mode)
    }
}

/// Lays out `[fallback, glyph₀, glyph₁, …]` as upright vertical lines,
/// or places `fallback` alone when the plan declines.
///
/// Subview 0 is ALWAYS the decline fallback and subviews 1…n are one glyph
/// each, in logical order — the same slot contract the Compose twin's
/// `Layout` content uses, so the two placement loops read alike.
struct VerticalUprightTextFlowLayout: Layout {
    /// The run as per-scalar strings — `VerticalTextFlow.codePointsOf(_:)`.
    /// Held (rather than recomputed) so the plan's indices and the subview
    /// order come from ONE split.
    let glyphs: [String]
    /// Which side line 1 sits on — `VerticalTextFlow.lineStack(_:)`.
    let stack: LineStack
    /// An explicit block-axis budget in px, or nil to read `proposal.height`.
    var budgetOverridePx: CGFloat? = nil

    /// The measured plan: per-line glyph indices plus every glyph's natural
    /// size. Nil = DECLINE (the caller's frozen subtree wins).
    private func plan(proposal: ProposedViewSize,
                      subviews: Subviews) -> (lines: [[Int]], sizes: [CGSize])? {
        guard subviews.count > 1 else { return nil }
        // EVERY subview is sized on BOTH paths — including the fallback the
        // plan path never places. Sizing costs one text measurement and paints
        // nothing (an unPLACED subview is not drawn), and it keeps the two
        // paths symmetric with the Compose twin's measure loop.
        _ = subviews[0].sizeThatFits(proposal)
        // Every glyph measures UNSPECIFIED: an upright glyph is its own line
        // box and is never squeezed by the run's box (CSS overflows instead).
        let sizes = (1..<subviews.count).map { subviews[$0].sizeThatFits(.unspecified) }
        // The advance along the vertical inline axis = one line box's height.
        let advance = Double(sizes.first?.height ?? 0)
        let raw = budgetOverridePx ?? proposal.height
        // `.infinity` is SwiftUI's "as much as you like" and is NOT a wrap
        // budget — map it to nil so the planner declines rather than packing
        // the whole run into one impossibly long line.
        let budget: Double? = raw.flatMap { $0.isFinite ? Double($0) : nil }
        guard let lines = VerticalTextFlow.uprightColumnIndices(glyphs: glyphs,
                                                               charAdvancePx: advance,
                                                               budgetPx: budget)
        else {
            // No silent fallthrough: the GATE already said this run is
            // upright, so a decline here is a real, named gap and not a
            // routine "not our case". Logged once per process.
            _ = PropertyTracker.logOnce(
                key: "writing-mode:upright-vertical-budget",
                message: "upright vertical run declined: no finite block-axis " +
                    "budget (proposal.height was \(String(describing: raw))). " +
                    "PlaceholderLabel's .fixedSize(vertical: true) erases that " +
                    "axis, so the wrap width needs a containing-block channel " +
                    "that publishes MAX-height bases; run kept horizontal.")
            return nil
        }
        return (lines, sizes)
    }

    /// Per-line cross extent (the line box's width) and main extent (the sum
    /// of its glyph advances) — shared by measure and place so the two can
    /// never compute different columns.
    private func extents(_ lines: [[Int]], _ sizes: [CGSize]) -> (w: [CGFloat], h: [CGFloat]) {
        let w = lines.map { line in line.map { sizes[$0].width }.max() ?? 0 }
        let h = lines.map { line in line.reduce(CGFloat(0)) { $0 + sizes[$1].height } }
        return (w, h)
    }

    func sizeThatFits(proposal: ProposedViewSize,
                      subviews: Subviews,
                      cache: inout Void) -> CGSize {
        guard let (lines, sizes) = plan(proposal: proposal, subviews: subviews) else {
            // DECLINE — the frame is whatever the frozen subtree wanted.
            return subviews.first?.sizeThatFits(proposal) ?? .zero
        }
        let (w, h) = extents(lines, sizes)
        return CGSize(width: w.reduce(0, +), height: h.max() ?? 0)
    }

    func placeSubviews(in bounds: CGRect,
                       proposal: ProposedViewSize,
                       subviews: Subviews,
                       cache: inout Void) {
        guard let (lines, sizes) = plan(proposal: proposal, subviews: subviews) else {
            subviews.first?.place(at: bounds.origin, anchor: .topLeading, proposal: proposal)
            return
        }
        let (lineW, _) = extents(lines, sizes)
        let totalW = lineW.reduce(0, +)
        // `vertical-rl` puts line 1 at the RIGHT edge and walks left;
        // `vertical-lr` starts at the left edge and walks right. Both walk the
        // plan in LOGICAL order — only the anchor differs.
        var x: CGFloat = (stack == .rightToLeft) ? totalW : 0
        for (index, line) in lines.enumerated() {
            if stack == .rightToLeft { x -= lineW[index] }
            var y: CGFloat = 0
            for glyphIndex in line {
                let size = sizes[glyphIndex]
                // Centre the glyph across its line box — the vertical
                // typesetting equivalent of a baseline-centred glyph in a
                // horizontal line box. A no-op when every glyph in the line
                // has the same advance (the CJK/fullwidth case, i.e. every
                // upright run).
                subviews[glyphIndex + 1].place(
                    at: CGPoint(x: bounds.minX + x + (lineW[index] - size.width) / 2,
                                y: bounds.minY + y),
                    anchor: .topLeading,
                    proposal: ProposedViewSize(size))
                y += size.height
            }
            if stack == .leftToRight { x += lineW[index] }
        }
    }
}

/// View wrapper for `VerticalUprightTextFlowLayout` — splits the run into
/// glyph slots once and hands the layout its `[fallback, glyphs…]` subviews.
struct VerticalUprightTextFlow<Fallback: View, Glyph: View>: View {
    /// The run as the caller will paint it (post text-transform).
    let text: String
    /// Which side line 1 sits on — `VerticalTextFlow.lineStack(_:)`.
    let stack: LineStack
    /// An explicit block-axis budget in px, or nil to read the proposal.
    var budgetOverridePx: CGFloat? = nil
    /// The frozen horizontal subtree, placed verbatim when the plan declines.
    @ViewBuilder var fallback: () -> Fallback
    /// Paints ONE code point upright, styled like the run it came from.
    @ViewBuilder var glyph: (String) -> Glyph

    var body: some View {
        let slots = VerticalTextFlow.codePointsOf(text)
        VerticalUprightTextFlowLayout(glyphs: slots,
                                      stack: stack,
                                      budgetOverridePx: budgetOverridePx) {
            fallback()                                   // subview 0
            // `id: \.offset` — the run may repeat a glyph, so the STRING is
            // not a stable identity; the slot index is.
            ForEach(Array(slots.enumerated()), id: \.offset) { _, s in
                glyph(s)                                 // subviews 1 … n
            }
        }
    }
}

File: runtimes/compose/src/main/java/com/styleconverter/runtime/core/renderer/VerticalTextFlowLayout.kt
package com.styleconverter.runtime.core.renderer

// Compose runtime — wave 35 lane B5, the renderer's VERTICAL-FLOW region.
//
// Two composables, both owned by `writing-mode` and nothing else. They are
// the ONLY place in the Compose renderer that turns a vertical writing mode
// into geometry; every decision they take comes from the pure twin module
// `typography/text/VerticalTextFlow.kt`, so the Kotlin and Swift renderers
// cannot drift about which runs are upright or where a vertical line breaks.
//
//   • [VerticalRotatedTextRun] — the wave-5 SIDEWAYS run, lifted verbatim out
//     of ComponentRenderer.PlaceholderContent. Measures the run against
//     SWAPPED constraints (its inline axis is the box's block axis) and spins
//     the laid-out result about its centre. Byte-identical to the code it
//     replaces: same swap, same coerce, same centre-rotate.
//   • [VerticalUprightTextFlow] — NEW. Typesets an UPRIGHT run as real
//     vertical lines: glyphs stack DOWN a line, lines stack across the block
//     axis (right-to-left for `vertical-rl`, left-to-right for `vertical-lr`).
//
// ## Why a second composable instead of a flag on the first
// A rotated run and an upright run need OPPOSITE transforms. The rotated path
// gets its line stacking for free — rotating a 4-line horizontal block by 90°
// turns the top line into the rightmost column — but every glyph comes along
// for the ride and ends up lying on its side. An upright run needs the line
// stacking WITHOUT the glyph rotation, which no single `rotationZ` can
// express. Measured on the frozen wave34-final capture of WPT
// css-writing-modes/available-size-011 (fullwidth "ＰＡＳＳ",
// Vertical_Orientation U ⇒ upright under the initial `text-orientation:
// mixed`): Android 0.9247 through the rotated path, web 0.9898 upright.
//
// ## The decline contract
// [VerticalUprightTextFlow] does its planning at MEASURE time, because the
// wrap budget is the incoming height constraint and composition cannot see
// it. When `VerticalTextFlow.uprightColumnIndices` declines — an unbounded
// height budget, a degenerate advance, a >64-line plan — this composable
// places the `rotatedRun` slot instead and the frame is exactly what the
// frozen renderer drew. Nothing is guessed.

import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.layout.Layout
import com.styleconverter.runtime.typography.text.GlyphOrientation
import com.styleconverter.runtime.typography.text.LineStack
import com.styleconverter.runtime.typography.text.VerticalTextFlow
import com.styleconverter.runtime.typography.text.WritingModeConfig

/**
 * THE GATE: does this run take the upright vertical path, and if so which way
 * do its lines stack? `null` = "no", and the caller must keep whatever it did
 * before — a horizontal mode, a sideways mode, an ASCII run under `mixed`, an
 * explicit `text-orientation: sideways`, or a run that needs both classes.
 *
 * One function so the decision has ONE spelling per platform; the Swift twin
 * is `VerticalUprightGate.stack(properties:text:)`, which reaches the same
 * two `VerticalTextFlow` calls through its own extractors.
 */
/** One-shot latch for the decline breadcrumb below (process-wide). */
private val uprightDeclineLogged = java.util.concurrent.atomic.AtomicBoolean(false)

internal fun verticalUprightStack(config: WritingModeConfig, text: String): LineStack? {
    val orientation = VerticalTextFlow.runOrientation(
        config.writingMode, config.textOrientation, text
    )
    if (orientation != GlyphOrientation.UPRIGHT) return null
    return VerticalTextFlow.lineStack(config.writingMode)
}

/**
 * The wave-5 SIDEWAYS vertical run: lay [run] out against swapped
 * constraints, then rotate it about its centre by [rotationDegrees].
 *
 * css-writing-modes-4 §3 changes the flow direction of the CONTENT, not the
 * box: the element keeps its specified width/height (that is why
 * `WritingModeApplier.applyWritingMode` returns the modifier untouched) and
 * only the glyph run turns. The run's inline axis is the box's BLOCK axis, so
 * it wraps at the incoming HEIGHT budget — hence the swap below.
 *
 * @param rotationDegrees +90 for the `-rl` family and `vertical-lr`
 *   (top-to-bottom glyph run), −90 for `sideways-lr` (which reads
 *   bottom-to-top). The caller owns that choice; this layout only applies it.
 */
@Composable
internal fun VerticalRotatedTextRun(
    rotationDegrees: Float,
    modifier: Modifier = Modifier,
    run: @Composable () -> Unit,
) {
    // Wave 48 (lane W1): the measure moved VERBATIM into
    // rotatedRunMeasurePolicy so the policy can carry EXPLICIT transposed
    // intrinsics. The trailing-lambda overload inherited the DEFAULT
    // intrinsics, whose fake placeables substitute LargeDimension (32767)
    // for an unbounded axis — and this layout's axis swap then reported
    // that sentinel as real perpendicular geometry (each rotated <td>
    // answered a 32769 max-content width, the table summed 98311, the
    // row's capped §17.5.3 height became 8190, and every vertical root
    // composed 8264 px tall — the direction-upright-002 unmeasured cell).
    // Full measured chain + the transposition table: VerticalRunIntrinsics.
    Layout(
        modifier = modifier,
        content = run,
        measurePolicy = remember(rotationDegrees) {
            rotatedRunMeasurePolicy(rotationDegrees)
        },
    )
}

/**
 * An UPRIGHT vertical run: glyphs stand up and stack DOWN each line, lines
 * stack across the block axis per [stack].
 *
 * The content carries BOTH candidate subtrees — slot 0 is [rotatedRun] (the
 * decline fallback), slots 1…n are one [uprightGlyph] per code point of
 * [text] — because the plan needs the incoming constraints and composition
 * runs before measurement. Exactly one subtree is ever PLACED, so only one
 * ever paints.
 *
 * TODO(wave-35 lane B5, honest gap): text-decoration lines and emphasis marks
 * are NOT carried onto an upright vertical run. The caller's `decorationModifier`
 * / `emphasisModifier` draw in the horizontal run's frame, and re-deriving
 * them per upright line box is a separate slice. No corpus test decorates an
 * upright vertical run today (the one that decorates a VERTICAL run,
 * css-text-decor/line-through-vertical, is ASCII ⇒ rotated ⇒ untouched by
 * this path).
 *
 * @param text          the run as the caller will paint it (post
 *   text-transform), split here into per-code-point glyph slots.
 * @param stack         which side line 1 sits on — `VerticalTextFlow.lineStack`.
 * @param rotatedRun    the frozen sideways path, placed verbatim on decline.
 * @param uprightGlyph  paints ONE code point upright, styled like the run.
 */
@Composable
internal fun VerticalUprightTextFlow(
    text: String,
    stack: LineStack,
    modifier: Modifier = Modifier,
    rotatedRun: @Composable () -> Unit,
    uprightGlyph: @Composable (String) -> Unit,
) {
    // One slot per code point, in logical order — the array the plan's
    // indices address. Memoised on the string so recomposition of an
    // unchanged run does not rebuild the list.
    val glyphs = remember(text) { VerticalTextFlow.codePointsOf(text) }
    // Wave 48 (lane W1): the measure moved VERBATIM into
    // uprightFlowMeasurePolicy so the policy carries EXPLICIT glyph-based
    // intrinsics — the default lambda-replay intrinsics were poisoned by
    // the same LargeDimension (32767) fakes as the rotated run's (this
    // flow's slot 0 IS a rotated run, and each glyph fake answered 32767
    // on its unbounded axis). See VerticalRunIntrinsics for the measured
    // chain and the css-sizing-3 §4 min/max-content model the overrides
    // implement. The decline breadcrumb stays here (the composable owns
    // the log channel; the policy stays log-free and JVM-testable).
    Layout(
        modifier = modifier,
        content = {
            rotatedRun()                          // slot 0 — decline fallback
            glyphs.forEach { uprightGlyph(it) }   // slots 1 … n
        },
        measurePolicy = remember(glyphs, stack) {
            uprightFlowMeasurePolicy(glyphs, stack, onDecline = {
                // No silent fallthrough: the GATE already said this run is
                // upright, so a decline is a real, named gap and not a
                // routine "not our case". Logged once per process — the
                // Swift twin logs the same key via PropertyTracker.logOnce.
                if (uprightDeclineLogged.compareAndSet(false, true)) {
                    android.util.Log.i(
                        "ComponentRenderer",
                        "writing-mode:upright-vertical-budget — upright vertical run " +
                            "declined: no finite block-axis budget; " +
                            "run kept on the rotated path",
                    )
                }
            })
        },
    )
}
